openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3364,319 @@ def from_openeye(oemol, allow_undefined_stereo=False):
+                    print('Invalid conformer for this molecule, the geometry could not be attached.')
+            # attach a dict that has the initial molecule ids and the number of the conformer it is stored in
+            offmol._properties['initial_molecules'] = initial_ids
+
+        return offmol
+
+    @classmethod
+    @RDKitToolkitWrapper.requires_toolkit()
+    def from_pdb_and_smiles(cls, file_path, smiles, allow_undefined_stereo=False):
+        """
+        Create a Molecule from a pdb file and a SMILES string using RDKit.
+
+        Requires RDKit to be installed.
+
+        The molecule is created and sanitised based on the SMILES string, we then find a mapping
+        between this molecule and one from the PDB based only on atomic number and connections.
+        The SMILES molecule is then reindex to match the PDB, the conformer is attached and the
+        molecule returned.
+
+        Parameters
+        ----------
+        file_path: str
+            PDB file path
+        smiles : str
+            a valid smiles string for the pdb, used for seterochemistry and bond order
+        allow_undefined_stereo : bool, default=False
+            If false, raises an exception if oemol contains undefined stereochemistry.
+
+        Returns
+        --------
+        molecule : openforcefield.Molecule
+            An OFFMol instance with ordering the same as used in the PDB file.
+
+        Raises
+        ------
+        InvalidConformerError : if the SMILES and PDB molecules are not isomorphic.
+        """
+
+        from rdkit import Chem
+
+        # Make the molecule from smiles
+        offmol = cls.from_smiles(smiles, allow_undefined_stereo=allow_undefined_stereo)
+
+        # Make another molecule from the PDB, allow stero errors here they are expected
+        pdbmol = cls.from_rdkit(Chem.MolFromPDBFile(file_path, removeHs=False), allow_undefined_stereo=True)
+
+        # check isomorphic and get the mapping if true the mapping will be
+        # Dict[pdb_index: offmol_index] sorted by pdb_index
+        isomorphic, mapping = Molecule.are_isomorphic(pdbmol, offmol, return_atom_map=True,
+                                                      aromatic_matching=False,
+                                                      formal_charge_matching=False,
+                                                      bond_order_matching=False,
+                                                      atom_stereochemistry_matching=False,
+                                                      bond_stereochemistry_matching=False)
+
+        if mapping is not None:
+            new_mol = offmol.remap(mapping)
+
+            # the pdb conformer is in the correct order so just attach it here
+            new_mol.add_conformer(pdbmol.conformers[0])
+
+            return new_mol
+
+        else:
+            raise InvalidConformerError('The PDB and SMILES structures do not match.')
+
+    def remap(self, mapping_dict, current_to_new=True):
```

Comment:
This function seems like a re-implementation of `_copy_initializer`, except that things are reordered here. I am concerned that it will be difficult to maintain these versions separately.

Reply:
This is a really good big-picture suggestion, and it's something that I'd totally passed over. I don't think we need to tackle it in this PR, but I'll create an Issue out of it.

Reply:
Good point attribute changes in future will mean that both functions need to be updated for now the testing of `remap` where the `dict()` of the atoms and bonds is compared after transferring attributes should catch where we introduce a new attribute and warn that the function is to be updated to support this.

Reply:
I have also changed this function to use the dict representation of atoms and bonds in line with `copy_initializer` to help with transferring all attributes.

Reply:
Ok, I've (finally) split this off into #496. Again, this is a good point, but I think this PR is creeping larger and larger, so I'd like to consolidate our gains before it gets too wide.